File: app/src/main/java/com/wmods/wppenhacer/xposed/features/general/CallType.java
package com.wmods.wppenhacer.xposed.features.general;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.Intent;
import android.net.Uri;
import android.os.BaseBundle;

import androidx.annotation.NonNull;

import com.wmods.wppenhacer.xposed.core.Feature;
import com.wmods.wppenhacer.xposed.core.WppCore;
import com.wmods.wppenhacer.xposed.core.components.AlertDialogWpp;
import com.wmods.wppenhacer.xposed.core.components.FMessageWpp;
import com.wmods.wppenhacer.xposed.core.devkit.Unobfuscator;
import com.wmods.wppenhacer.xposed.core.devkit.UnobfuscatorCache;
import com.wmods.wppenhacer.xposed.utils.ReflectionUtils;
import com.wmods.wppenhacer.xposed.utils.ResId;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XSharedPreferences;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;

public class CallType extends Feature {
    private XC_MethodHook.Unhook hookBundleBoolean;

    public CallType(@NonNull ClassLoader loader, @NonNull XSharedPreferences preferences) {
        super(loader, preferences);
    }

    @SuppressLint("ApplySharedPref")
    @Override
    public void doHook() throws Throwable {

        if (!prefs.getBoolean("calltype", false)) return;

        var intPreferences = Unobfuscator.loadGetIntPreferences(classLoader);
        XposedBridge.hookMethod(intPreferences, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                if (param.args[1] == "call_confirmation_dialog_count") {
                    param.setResult(1);
                }
            }
        });


        var callConfirmationFragment = XposedHelpers.findClass("com.whatsapp.calling.fragment.CallConfirmationFragment", classLoader);
        var method = ReflectionUtils.findMethodUsingFilter(callConfirmationFragment, m -> m.getParameterCount() == 1 && m.getParameterTypes()[0].equals(android.os.Bundle.class));
        XposedBridge.hookMethod(method, new XC_MethodHook() {
            private boolean isVideoCall;
            private String jid;
            private Dialog newDialog;
            private Unhook hookBundleString;

            @Override
            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                hookBundleString = XposedHelpers.findAndHookMethod(BaseBundle.class, "getString", String.class, new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                        if (param.args[0] == "jid") {
                            jid = (String) param.getResult();
                        }
                    }
                });
                hookBundleBoolean = XposedHelpers.findAndHookMethod(BaseBundle.class, "getBoolean", String.class, new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                        if (param.args[0] == "is_video_call") {
                            isVideoCall = (boolean) param.getResult();
                        }
                    }
                });
            }

            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                hookBundleString.unhook();
                hookBundleBoolean.unhook();
                if (jid == null || isVideoCall) return;
                var origDialog = (Dialog) param.getResult();
                var context = origDialog.getContext();
                var mAlertDialog = new AlertDialogWpp(origDialog.getContext());
                mAlertDialog.setTitle(UnobfuscatorCache.getInstance().getString("selectcalltype"));
                mAlertDialog.setItems(new String[]{context.getString(ResId.string.phone_call), context.getString(ResId.string.whatsapp_call)}, (dialog, which) -> {
                    newDialog.dismiss();
                    switch (which) {
                        case 0:
                            var intent = new Intent();
                            intent.setAction(Intent.ACTION_DIAL);
                            var userJid = new FMessageWpp.UserJid(jid);
                            intent.setData(Uri.parse("tel:+" + userJid.getStripJID()));
                            context.startActivity(intent);
                            break;
                        case 1:
                            origDialog.show();
                            break;
                    }
                });
                newDialog = mAlertDialog.create();
                param.setResult(newDialog);
            }
        });
    }

    @NonNull
    @Override
    public String getPluginName() {
        return "Call Type";
    }
}

File: app/src/main/java/com/wmods/wppenhacer/xposed/features/general/CallType.kt
package com.wmods.wppenhacer.xposed.features.general

import android.app.Dialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import androidx.core.net.toUri
import com.wmods.wppenhacer.R
import com.wmods.wppenhacer.xposed.core.Feature
import com.wmods.wppenhacer.xposed.core.components.AlertDialogWpp
import com.wmods.wppenhacer.xposed.core.components.FMessageWpp.UserJid
import com.wmods.wppenhacer.xposed.core.components.SharedPreferencesWrapper
import com.wmods.wppenhacer.xposed.core.devkit.UnobfuscatorCache
import com.wmods.wppenhacer.xposed.utils.ReflectionUtils
import de.robv.android.xposed.XC_MethodHook
import android.content.SharedPreferences 
import de.robv.android.xposed.XposedBridge
import de.robv.android.xposed.XposedHelpers

class CallType(loader: ClassLoader, preferences:SharedPreferences) :
    Feature(loader, preferences) {
    override fun doHook() {
        if (!prefs.getBoolean("calltype", false)) return

        SharedPreferencesWrapper.addHook { key, value ->
            if (key == "call_confirmation_dialog_count") {
                return@addHook 1
            }
            value
        }


        val callConfirmationFragment = XposedHelpers.findClass(
            "com.whatsapp.calling.fragment.CallConfirmationFragment",
            classLoader
        )
        val method = ReflectionUtils.findMethodUsingFilter(
            callConfirmationFragment
        ) { m -> m.parameterCount == 1 && m.parameterTypes[0] == Bundle::class.java }
        XposedBridge.hookMethod(method, object : XC_MethodHook() {
            override fun beforeHookedMethod(param: MethodHookParam) {
                val bundle = param.args.firstOrNull() as? Bundle ?: return
                param.setObjectExtra("wae_call_jid", bundle.getString("jid"))
                param.setObjectExtra("wae_is_video_call", bundle.getBoolean("is_video_call"))
            }

            @Throws(Throwable::class)
            override fun afterHookedMethod(param: MethodHookParam) {
                val jid = param.getObjectExtra("wae_call_jid") as? String
                val isVideoCall = param.getObjectExtra("wae_is_video_call") as? Boolean ?: false
                if (jid == null || isVideoCall) return
                val origDialog = param.result as Dialog
                val context = origDialog.context
                val mAlertDialog = AlertDialogWpp(origDialog.context)
                lateinit var newDialog: Dialog
                mAlertDialog.setTitle(UnobfuscatorCache.getInstance().getString("selectcalltype"))
                mAlertDialog.setItems(
                    arrayOf(
                        context.getString(R.string.phone_call),
                        context.getString(R.string.whatsapp_call)
                    )
                ) { _: DialogInterface?, which: Int ->
                    newDialog!!.dismiss()
                    when (which) {
                        0 -> {
                            val intent = Intent()
                            intent.action = Intent.ACTION_DIAL
                            val userJid = UserJid(jid)
                            intent.data = ("tel:+" + userJid.phoneNumber).toUri()
                            context.startActivity(intent)
                        }

                        1 -> origDialog.show()
                    }
                }
                newDialog = mAlertDialog.create()
                param.setResult(newDialog)
            }
        })
    }

    override fun getPluginName(): String {
        return "Call Type"
    }
}
